File: JavaScript/42-Trapping Rain Water.js
/**
 * @param {number[]} height
 * @return {number}
 */
var trap = function(height) {
    let res = 0;
    let stack = [];
    for (let i=0; i<height.length; i++){
        while (stack.length>0 && height[i] > height[stack[stack.length-1]]){
            let currIndex = stack.pop();
            if (stack.length>0){
                let h = Math.min(height[i], height[stack[stack.length-1]]) - height[currIndex];
                let w = i - stack[stack.length-1] - 1;
                res += h*w
            }
        }
        stack.push(i);
    }
    return res;
};

File: C++/42-Trapping Rain Water.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        stack<int> stk;
        int res = 0;
        for (int i=0; i<height.size(); i++) {
            while (!stk.empty() && height[i] > height[stk.top()]) {
                int prevH = height[stk.top()];
                stk.pop();
                if (stk.empty()) break; 
                int w = i - stk.top() - 1;
                int h = min(height[i], height[stk.top()]) - prevH;
                res += w*h;
            }
            stk.push(i);
        }
        return res;
    }
};